client/client-lib/src/api.rs

```diff
@@ -1,148 +1,615 @@
 use fedimint_core::modules::ln::contracts::incoming::IncomingContractOffer;
 use fedimint_core::modules::ln::contracts::ContractId;
 use fedimint_core::modules::ln::{ContractAccount, LightningGateway};
-use fedimint_core::modules::mint::db::ECashUserBackupSnapshot;
 use fedimint_core::modules::wallet::PegOutFees;
-use fedimint_core::outcome::legacy::{OutputOutcome, TryIntoOutcome};
-use fedimint_core::outcome::TransactionStatus;
-use fedimint_core::transaction::legacy::Transaction as LegacyTransaction;
+use fedimint_core::outcome::legacy::TryIntoOutcome;
+use fedimint_core::outcome::{self, TransactionStatus};
 use fedimint_core::transaction::SerdeTransaction;
 use fedimint_core::CoreError;
+use fedimint_mint::db::ECashUserBackupSnapshot;
 use futures::stream::FuturesUnordered;
-use futures::StreamExt;
+use futures::{Future, StreamExt};
 #[cfg(not(target_family = "wasm"))]
 use jsonrpsee_core::client::CertificateStore;
 use jsonrpsee_core::client::ClientT;
 use jsonrpsee_core::Error as JsonRpcError;
-use jsonrpsee_types::error::CallError as RpcCallError;
 #[cfg(target_family = "wasm")]
 use jsonrpsee_wasm_client::{Client as WsClient, WasmClientBuilder as WsClientBuilder};
 #[cfg(not(target_family = "wasm"))]
 use jsonrpsee_ws_client::{WsClient, WsClientBuilder};
 use serde::{Deserialize, Serialize};
+use serde_json::Value;
 use thiserror::Error;
 use threshold_crypto::PublicKey;
 use tracing::{debug, error, instrument, trace, warn};
 use url::Url;
 
-use crate::module_decode_stubs;
 use crate::query::{
     CurrentConsensus, EventuallyConsistent, QueryStep, QueryStrategy, Retry404, UnionResponses,
     UnionResponsesSingle, ValidHistory,
 };
+use crate::LegacyTransaction;
+
+pub type Result<T> = result::Result<T, ApiError>;
+
+pub type JsonRpcResult<T> = result::Result<T, jsonrpsee_core::Error>;
+pub type FedResult<T> = result::Result<T, FedApiError>;
+
+pub mod fake;
+
+// TODO: rename to `MemberApiError`
+#[derive(Debug, Error)]
+pub enum ApiError {
+    // TODO: this sticks out like a sore thumb
+    // #[error("Params serialization error: {0}")]
+    // ParamsSerialization(anyhow::Error),
+    #[error("Response deserialization error: {0}")]
+    ResponseDeserialization(anyhow::Error),
+    #[error("Invalid peer id: {peer_id}")]
+    InvalidPeerId { peer_id: PeerId },
+    #[error("Rpc error: {0}")]
+    Rpc(#[from] JsonRpcError),
+}
+
+impl ApiError {
+    pub fn is_retryable(&self) -> bool {
+        match self {
+            ApiError::ResponseDeserialization(_) => false,
+            ApiError::InvalidPeerId { peer_id: _ } => false,
+            ApiError::Rpc(rpc_e) => match rpc_e {
+                JsonRpcError::Transport(_) => true,
+                JsonRpcError::Internal(_) => true,
+                JsonRpcError::Call(jsonrpsee_types::error::CallError::Custom(e)) => e.code() == 404,
+                _ => false,
+            },
+        }
+    }
+}
+
+#[derive(Debug, Error)]
+pub struct FedApiError(BTreeMap<PeerId, ApiError>);
```

Comment:
Neat part: now that we have a different member vs federation api error, we will report responses from all peers if the call fail etc. Should help the debugging.

Reply:
That's great